src/gnutls_creds.cpp

```diff
@@ -0,0 +1,277 @@
+            throw std::runtime_error("gnutls import of raw Ed keys failed");
+        }
+
+        // clang format keeps changing this arbitrarily, so disable for this line
+        // clang-format off
+        constexpr auto* priority = "NORMAL:+ECDHE-PSK:+PSK:+ECDHE-ECDSA:+AES-128-CCM-8:+CTYPE-CLI-ALL:+CTYPE-SRV-ALL:+SHA256";
+        // clang-format on
+
+        const char* err{nullptr};
+        if (auto rv = gnutls_priority_init(&priority_cache, priority, &err); rv < 0)
+        {
+            if (rv == GNUTLS_E_INVALID_REQUEST)
+                log::error(log_cat, "gnutls_priority_init error: {}", err);
+            else
+                log::error(log_cat, "gnutls_priority_init error: {}", gnutls_strerror(rv));
+
+            throw std::runtime_error("gnutls key exchange algorithm priority setup failed");
+        }
+
+        // NOTE: the original place this was called
+        // gnutls_certificate_set_verify_function(cred, cert_verify_callback_gnutls);
+    }
+
+    GNUTLSCreds::~GNUTLSCreds()
+    {
+        log::info(log_cat, "Entered {}", __PRETTY_FUNCTION__);
+        gnutls_certificate_free_credentials(cred);
+    }
+
+    std::shared_ptr<GNUTLSCreds> GNUTLSCreds::make(
+            std::string remote_key, std::string remote_cert, std::string local_cert, std::string ca)
+    {
+        // would use make_shared, but I want GNUTLSCreds' constructor to be private
+        std::shared_ptr<GNUTLSCreds> p{new GNUTLSCreds(remote_key, remote_cert, local_cert, ca)};
+        return p;
+    }
+
+    std::shared_ptr<GNUTLSCreds> GNUTLSCreds::make_from_ed_keys(std::string seed, std::string pubkey)
+    {
+        // would use make_shared, but I want GNUTLSCreds' constructor to be private
+        std::shared_ptr<GNUTLSCreds> p{new GNUTLSCreds(seed, pubkey)};
+        return p;
+    }
+
+    std::shared_ptr<GNUTLSCreds> GNUTLSCreds::make_from_ed_seckey(std::string sk)
+    {
+        if (sk.size() != GNUTLS_SECRET_KEY_SIZE)
+            throw std::invalid_argument("Ed25519 secret key is invalid length!");
+
+        auto pk = sk.substr(GNUTLS_KEY_SIZE);
+        sk = sk.substr(0, GNUTLS_KEY_SIZE);
+
+        std::shared_ptr<GNUTLSCreds> p{new GNUTLSCreds(std::move(sk), std::move(pk))};
+        return p;
+    }
+
+    std::unique_ptr<TLSSession> GNUTLSCreds::make_session(bool is_client, const std::vector<std::string>& alpns)
+    {
+        // NOTE: I put this here (rather than just the raw_pk constructor) to test the server requesting certs every time,
+        // but not requiring them. By doing the check here, we can ensure that requesting is only done by servers, while
+        // retrieval is only done by clients
+        if (using_raw_pk)
+        {
+            // if (is_client)
+            //     gnutls_certificate_set_retrieve_function2(cred, cert_retrieve_callback_gnutls);
+            // else
+            gnutls_certificate_set_verify_function(cred, cert_verify_callback_gnutls);
```

Comment:
This feels a little akward because we'll be setting this on `creds` again and again for every session we create.  Reading the comment above, I'm assuming this is just for the test suite, so I'm wondering if we can find a way to avoid the awkwardness?

Reply:
Moved into the constructor